File: app/src/test/java/test/notificationreader/model/notifications/NotificationFactoryTests.java
package test.notificationreader.model.notifications;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import test.notificationreader.model.TextReader;

import static org.assertj.core.api.Assertions.assertThat;


public class NotificationFactoryTests {

    @Mock
    TextReader mTextReader;
    NotificationActor mActor;

    @Before
    public void setupMocks() {
        MockitoAnnotations.initMocks(this);
        mActor = new NotificationActor(mTextReader);
    }

    @Test
    public void buildMessageNotification() {
        Notification notification = NotificationFactory.INSTANCE.build("Maria: hello Daniel", "facebook.messenger");

        assertThat(notification).isInstanceOf(MessageNotification.class);


        MessageNotification messageNotification = (MessageNotification) notification;
        assertThat(messageNotification.getName()).isEqualTo("Maria");
        assertThat(messageNotification.getMessage()).isEqualTo("hello Daniel");
    }

    @Test
    public void buildSystemNotification() {
        Notification notification = NotificationFactory.INSTANCE.build("Something happened with phone", "android.vending");
        assertThat(notification).isInstanceOf(SystemNotification.class);
    }

    @Test
    public void buildOtherNotification() {
        Notification notification = NotificationFactory.INSTANCE.build("Maria hello Daniel", "facebook.messenger");

        assertThat(notification).isInstanceOf(Notification.class);
        assertThat(notification).isNotInstanceOf(MessageNotification.class);
        assertThat(notification).isNotInstanceOf(SystemNotification.class);
    }

}

File: app/src/test/java/test/notificationreader/model/notifications/NotificationFactoryTests.kt
package test.notificationreader.model.notifications

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class NotificationFactoryTests {
    @Test
    fun buildMessageNotification() {
        val notification = NotificationFactory.build("Maria: hello Daniel", "facebook.messenger")

        assertThat(notification).isInstanceOf(MessageNotification::class.java)

        val messageNotification = notification as MessageNotification
        assertThat(messageNotification.name).isEqualTo("Maria")
        assertThat(messageNotification.message).isEqualTo("hello Daniel")
    }

    @Test
    fun buildSystemNotification() {
        val notification = NotificationFactory.build("Something happened with phone", "android.vending")
        assertThat(notification).isInstanceOf(SystemNotification::class.java)
    }

    @Test
    fun buildOtherNotification() {
        val notification = NotificationFactory.build("Maria hello Daniel", "facebook.messenger")

        assertThat(notification).isInstanceOf(Notification::class.java)
        assertThat(notification).isNotInstanceOf(MessageNotification::class.java)
        assertThat(notification).isNotInstanceOf(SystemNotification::class.java)
    }
}
